File: mcd_dart/lib/core/recognition.dart
import 'dart:math' as math;
import 'package:image/image.dart';

import '../models/image.dart';
import '../utils/image_hash.dart';

class RecognitionResult {
  final bool isRecognized;
  final double recognitionScore;
  final String cardName;
  
  RecognitionResult(this.isRecognized, this.recognitionScore, this.cardName);
}

List<double> phashDiff(List<ReferenceImage> referenceImages, ImageHash phashIm) {
  // Calculate the difference between the given hash and each reference image hash
  List<double> diff = List.filled(referenceImages.length, 0.0);
  
  for (int i = 0; i < referenceImages.length; i++) {
    // Calculate the hash distance
    if (referenceImages[i].phash != null) {
      diff[i] = phashIm.distanceTo(referenceImages[i].phash!);
    } else {
      diff[i] = double.infinity;
    }
  }
  
  return diff;
}

RecognitionResult phashCompare(
  Image imSeg, 
  List<ReferenceImage> referenceImages, 
  {double hashSeparationThr = 4.0, 
  bool verbose = false}
) {
  String cardName = 'unknown';
  bool isRecognized = false;
  double recognitionScore = 0.0;
  
  // Try different rotations
  List<double> rotations = [0.0, 90.0, 180.0, 270.0];
  
  List<double> d0Dist = List.filled(rotations.length, 0.0);
  List<List<double>> d0 = List.generate(
    referenceImages.length, 
    (_) => List.filled(rotations.length, 0.0)
  );
  
  for (int j = 0; j < rotations.length; j++) {
    // Apply rotation to the image
    Image rotatedImage;
    if (rotations[j].abs() > 1e-5) {
      rotatedImage = copyRotate(imSeg, angle: rotations[j].toInt());
    } else {
      rotatedImage = imSeg.clone();
    }
    
    // Calculate perceptual hash
    ImageHash phashIm = ImageHash.createPerceptualHash(rotatedImage);
    
    // Calculate differences with all reference images
    d0.asMap().forEach((i, row) {
      row[j] = phashDiff(referenceImages, phashIm)[i];
    });
    
    // Calculate statistical distance
    double minDiff = double.infinity;
    for (var row in d0) {
      if (row[j] < minDiff) {
        minDiff = row[j];
      }
    }
    
    // Create a list of differences excluding the minimum
    List<double> d0Filtered = [];
    for (var row in d0) {
      if (row[j] > minDiff) {
        d0Filtered.add(row[j]);
      }
    }
    
    // Calculate average and standard deviation
    double d0Ave = calculateAverage(d0Filtered);
    double d0Std = calculateStdDev(d0Filtered, d0Ave);
    
    // Calculate statistical distance
    d0Dist[j] = (d0Std > 0) ? (d0Ave - minDiff) / d0Std : 0;
    
    if (verbose) {
      print('Phash statistical distance: ${d0Dist[j]}');
    }
    
    // Check if this rotation gives the best recognition result
    if (d0Dist[j] > hashSeparationThr && 
        d0Dist.indexOf(d0Dist.reduce(math.max)) == j) {
      
      // Find index of minimum difference
      int minIndex = 0;
      double minVal = double.infinity;
      
      for (int i = 0; i < d0.length; i++) {
        if (d0[i][j] < minVal) {
          minVal = d0[i][j];
          minIndex = i;
        }
      }
      
      // Get card name from reference image
      String fullName = referenceImages[minIndex].name;
      cardName = fullName.contains('.jpg') 
          ? fullName.split('.jpg')[0] 
          : fullName;
      
      isRecognized = true;
      recognitionScore = d0Dist[j] / hashSeparationThr;
      break;
    }
  }
  
  return RecognitionResult(isRecognized, recognitionScore, cardName);
}

// Helper statistical functions
double calculateAverage(List<double> values) {
  if (values.isEmpty) return 0;
  return values.reduce((a, b) => a + b) / values.length;
}

double calculateStdDev(List<double> values, double average) {
  if (values.isEmpty) return 0;
  double sumSquaredDiff = values.fold(0, (sum, val) => sum + math.pow(val - average, 2));
  return math.sqrt(sumSquaredDiff / values.length);
}

File: old/lib/core/recognition.py
"""
Card recognition functionality for MTG card detector.
"""

import numpy as np
import imagehash
from PIL import Image as PILImage
import cv2
from scipy.ndimage import rotate


def phash_diff(reference_images, phash_im):
    """
    Calculates the phash difference between the given phash and
    each of the reference images.
    """
    diff = np.zeros(len(reference_images))
    for i, ref_im in enumerate(reference_images):
        diff[i] = phash_im - ref_im.phash
    return diff


def phash_compare(im_seg, reference_images, hash_separation_thr=4.0, verbose=False):
    """
    Runs perceptive hash comparison between given image and
    the (pre-hashed) reference set.
    """

    card_name = 'unknown'
    is_recognized = False
    recognition_score = 0.
    rotations = np.array([0., 90., 180., 270.])

    d_0_dist = np.zeros(len(rotations))
    d_0 = np.zeros((len(reference_images), len(rotations)))
    for j, rot in enumerate(rotations):
        if not -1.e-5 < rot < 1.e-5:
            phash_im = imagehash.phash(
                PILImage.fromarray(np.uint8(255 * cv2.cvtColor(
                    rotate(im_seg, rot), cv2.COLOR_BGR2RGB))),
                hash_size=32)
        else:
            phash_im = imagehash.phash(
                PILImage.fromarray(np.uint8(255 * cv2.cvtColor(
                    im_seg, cv2.COLOR_BGR2RGB))),
                hash_size=32)
        d_0[:, j] = phash_diff(reference_images, phash_im)
        d_0_ = d_0[d_0[:, j] > np.amin(d_0[:, j]), j]
        d_0_ave = np.average(d_0_)
        d_0_std = np.std(d_0_)
        d_0_dist[j] = (d_0_ave - np.amin(d_0[:, j])) / d_0_std
        if verbose:
            print('Phash statistical distance: ' + str(d_0_dist[j]))
        if (d_0_dist[j] > hash_separation_thr and
                np.argmax(d_0_dist) == j):
            card_name = reference_images[np.argmin(d_0[:, j])]\
                .name.split('.jpg')[0]
            is_recognized = True
            recognition_score = d_0_dist[j] / hash_separation_thr
            break
    return (is_recognized, recognition_score, card_name)